File: gpu/GPUTutorial/redux.c
/*--------------------------------------------------------------------*/
/* C GPU Tutorial: Reduction        */
/* written by Viktor K. Decyk, UCLA */

#include "redux.h"

void sum0(float a[], float *sa, int nx) {
/* simple 1d sum reduction of length nx */
/* sa = sum(a)                           */
/* local data */
   int j;

   *sa = 0.0;
   for (j = 0; j < nx; j++) {
      *sa += a[j];
   }

   return;
}

void sum1(float a[], float *sa, int mx, int nx) {
/* 1d sum reductions, each of length mx */
/* sa = sum(a) */
/* local data */
   int j, js, jb, nbx, mxm;
   float t;
/* nbx = number of blocks */
   nbx = (nx - 1)/mx + 1;

   *sa = 0.0;
   for (jb = 0; jb < nbx; jb++) {
      mxm = mx < nx-mx*jb ? mx : nx-mx*jb;
      t = 0.0;
      for (js = 0; js < mxm; js++) {
         j = js + mx*jb;
         t += a[j];
      }
      *sa += t;
   }

   return;
}

void sum2(float a[], float d[], int mx, int nx) {
/* segmented 1d sum reductions, each of length mx             */
/* forall (j = 1:nbx); d(j) = sum(a(1+mx*(j-1):min(nx,mx*j))) */
/* local data */
   int j, js, jb, nbx, mxm;
   float t;
/* nbx = number of blocks */
   nbx = (nx - 1)/mx + 1;

   for (jb = 0; jb < nbx; jb++) {
      mxm = mx < nx-mx*jb ? mx : nx-mx*jb;
      t = 0.0;
      for (js = 0; js < mxm; js++) {
         j = js + mx*jb;
         t += a[j];
      }
      d[jb] = t;
   }

   return;
}

/* Interfaces to Fortran */

void csum0_(float *a, float *sa, int *nx) {
   sum0(a,sa,*nx);
   return;
}

void csum1_(float *a, float *sa, int *mx, int *nx) {
   sum1(a,sa,*mx,*nx);
   return;
}

void csum2_(float *a, float *d, int *mx, int *nx) {
   sum2(a,d,*mx,*nx);
   return;
}

File: gpu/GPUTutorial/redux.f90
!-----------------------------------------------------------------------
! Fortran GPU Tutorial: Reduction
! written by Viktor K. Decyk, UCLA
      module redux
      implicit none
!
      contains
!
      subroutine sum0(a,sa)
! simple 1d sum reduction of length nx
! sa = sum(a)
      implicit none
      real :: sa
      real, dimension(:) :: a
! local data
      integer :: j
!
      sa = 0.0
      do j = 1, size(a,1)
         sa = sa + a(j)
      enddo
!
      end subroutine
!
      subroutine sum1(a,sa,mx)
! 1d sum reductions, each of length mx
! sa = sum(a)
      implicit none
      integer :: mx
      real :: sa
      real, dimension(:) :: a
! local data
      integer :: j, js, jb, nx, nbx
      real :: t
! nx = size of a array
      nx = size(a,1)
! nbx = number of blocks
      nbx = (nx - 1)/mx + 1
!
      sa = 0.0
      do jb = 1, nbx
      t = 0.0
      do js = 1, min(mx,nx-mx*(jb-1))
         j = js + mx*(jb - 1)
         t = t + a(j)
      enddo
      sa = sa + t

      enddo
!
      end subroutine
!
      subroutine sum2(a,d,mx)
! segmented 1d sum reductions, each of length mx
! forall (j = 1:nbx); d(j) = sum(a(1+mx*(j-1):min(nx,mx*j))); end forall
      implicit none
      integer :: mx
      real, dimension(:) :: a, d
! local data
      integer :: j, js, jb, nx, nbx
      real :: t
! nx = size of a array
      nx = size(a,1)
! nbx = number of blocks
      nbx = (nx - 1)/mx + 1
!
      do jb = 1, nbx
      t = 0.0
      do js = 1, min(mx,nx-mx*(jb-1))
         j = js + mx*(jb - 1)
         t = t + a(j)
      enddo
      d(jb) = t
      enddo
!
      end subroutine
!
      end module

